src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java

```diff
@@ -686,10 +653,197 @@ public void afterIndexShardStarted(IndexShard indexShard) {
+        }
+
+        /**
+         * Requests an async configuration reload for the given configuration types. Calling this method
+         * will not wait for the configuration reload to complete.
+         */
+        void requestReload(Collection<CType<?>> configurationTypes, ActionListener<ConfigReloadResponse> actionListener) {
+            synchronized (this.requestLock) {
+                if (!this.started) {
+                    LOGGER.info("Cannot reload configuration yet, because the initialization process did not complete yet");
+                }
+
+                if (actionListener != null) {
+                    this.reloadRequestedForActionListeners.add(actionListener);
+                }
+
+                if (this.reloadRequestedFor.isEmpty()) {
+                    LOGGER.debug("Configuration reload request received for {}; notifying update thread", configurationTypes);
+                    this.reloadRequestedAt = Instant.now();
+                    this.reloadRequestedFor = ImmutableSet.copyOf(configurationTypes);
+                    this.requestLock.notifyAll();
+                } else if (!this.reloadRequestedFor.containsAll(configurationTypes)) {
+                    LOGGER.debug(
+                        "Configuration reload request received for {}; adding new configuration types to already requested {}",
+                        configurationTypes,
+                        this.reloadRequestedFor
+                    );
+                    this.reloadRequestedFor = ImmutableSet.<CType<?>>builder()
+                        .addAll(this.reloadRequestedFor)
+                        .addAll(configurationTypes)
+                        .build();
+                } else {
+                    if (Duration.between(this.reloadRequestedAt, Instant.now()).toMillis() > 30000) {
+                        // Reload request is queued for more than 30 seconds; let us log a warning about that
+                        LOGGER.warn(
+                            "Configuration reload request received; another update request is already queued since {}",
+                            this.reloadRequestedAt
+                        );
+                    } else {
+                        LOGGER.debug(
+                            "Configuration reload request received; another update request is already queued since {}",
+                            this.reloadRequestedAt
+                        );
+                    }
+                }
+            }
+        }
+
+        /**
+         * Starts the reload thread. Calling this method after the thread was already started will have no further effect.
+         */
+        void start() {
+            synchronized (this.requestLock) {
+                if (!this.started) {
+                    this.thread.start();
+                    this.started = true;
+                }
+            }
+        }
```

Comment:
Out of curiosity, will the thread ever be killed and re-created?

Reply:
No. It will be only killed when the JVM shuts down.